rm-config/src/keymap.rs

```diff
@@ -0,0 +1,425 @@
+                let action = action.ok_or_else(|| de::Error::missing_field("action"))??;
+                let modifier = modifier.transpose().unwrap();
+
+                if modifier.is_some() {
+                    if let KeyCode::Char(char) = on {
+                        if char.is_uppercase() {
+                            return Err(de::Error::custom(
+                                "you can't have a modifier with an uppercase letter, sorry",
+                            ));
+                        }
+                    }
+                }
+
+                Ok(Keybinding::new(on, action, modifier))
+            }
+        }
+
+        const FIELDS: &[&str] = &["on", "modifier", "action"];
+        deserializer.deserialize_struct(
+            "Keybinding",
+            FIELDS,
+            KeybindingVisitor {
+                phantom: PhantomData::default(),
+            },
+        )
+    }
+}
+
+#[derive(Serialize, Deserialize, Hash, Clone, Copy, PartialEq, Eq)]
+pub enum KeyModifier {
+    None,
+    Ctrl,
+    Shift,
+    Alt,
+    Super,
+    Meta,
+}
+
+impl KeyModifier {
+    fn to_str(self) -> &'static str {
+        match self {
+            KeyModifier::None => "",
+            KeyModifier::Ctrl => "CTRL",
+            KeyModifier::Shift => "SHIFT",
+            KeyModifier::Alt => "ALT",
+            KeyModifier::Super => "SUPER",
+            KeyModifier::Meta => "META",
+        }
+    }
+
+    fn is_none(self) -> bool {
+        self == KeyModifier::None
+    }
+}
+
+impl From<KeyModifier> for KeyModifiers {
+    fn from(value: KeyModifier) -> Self {
+        match value {
+            KeyModifier::None => KeyModifiers::NONE,
+            KeyModifier::Ctrl => KeyModifiers::CONTROL,
+            KeyModifier::Shift => KeyModifiers::SHIFT,
+            KeyModifier::Alt => KeyModifiers::ALT,
+            KeyModifier::Super => KeyModifiers::SUPER,
+            KeyModifier::Meta => KeyModifiers::META,
+        }
+    }
+}
+
+impl Default for KeyModifier {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+impl KeymapConfig {
+    pub const FILENAME: &'static str = "keymap.toml";
+    const DEFAULT_CONFIG: &'static str = include_str!("../defaults/keymap.toml");
+
+    pub fn init() -> Result<Self> {
+        match utils::fetch_config::<Self>(Self::FILENAME) {
+            Ok(mut keymap_config) => {
+                keymap_config.populate_hashmap();
+                return Ok(keymap_config);
+            }
+            Err(e) => match e {
+                utils::ConfigFetchingError::Io(e) if e.kind() == ErrorKind::NotFound => {
+                    let mut keymap_config =
+                        utils::put_config::<Self>(Self::DEFAULT_CONFIG, Self::FILENAME)?;
+                    keymap_config.populate_hashmap();
+                    return Ok(keymap_config);
+                }
+                utils::ConfigFetchingError::Toml(_) => anyhow::bail!(e),
+                _ => anyhow::bail!(e),
```

Comment:
is the `utils::ConfigFetchingError` needed? since both are bailed .. unless we plan on handling it differently in the future, if so maybe can add a TODO?

Reply:
There's still io::Error that needs to be checked whether or not it's NotFound (in this case a default config is being placed). But those 2 match branches can be merged into one